### ISSUE TITLE
Change wheel name

### ISSUE THREAD

**Issue by Hywan** [PyO3/maturin#313]
Hello,

I use Pyo3, thanks for the great software! I want to know whether it is possible to change the wheel name. I understand the wheel name is based on the crate name defined in the `Cargo.toml` file. However, I want to generate multiple wheels (depending of some Cargo features), like `myproject`, `myproject-xxx`, `myproject-yyy` etc. Is it possible to change that from the `maturin` commands, or do I have to open a PR?

**Comment by hombit** [comment]
I believe that wheel name is based on `Cargo.toml`'s `lib.name` but I'm fully support this possibility, it would be nice to have possibility to build and publish different Python packages for different `Cargo.toml`'s features

**Comment by konstin** [comment]
The wheel name is indeed based on `lib.name`. This is because the wheel filename should match the package/module name, which needs to match the name of your `pymodule` or python package directory. Pip also expects that, e.g. when using `pip install myproject --find-links /vendored/wheels/dir` it looks for `/vendored/wheels/dir/myproject-<tags>.whl` <sup>1</sup> .

A simple workaround would be creating a new crate in the workspace that has a unique name and only pulls in in the actual implementation crate with the desired features and exposes a `pymodule` with the correct name. You can then publish this as separate python package without any name clashes.

<sup>1</sup>There are exceptions such as PIL/Pillow, but imo that's against the intention of PEP 427 and leads to really nasty bugs, e.g. installing both `myproject-xxx` and `myproject-yyy` would give different results depending on the installation order.

**Comment by hombit** [comment]
I also found that wheel name (packages name and name I will use to `pip install` my package) is a crate name, while `lib.name` is a module name (name I will use to `import` my package). It could be useful to override wheel name with something. For example, I have some multi-crate repository, where my workspace contains crates `myproject` and `myproject-python`. In this case my pypi package will be named `myproject-python` when I it would much better to call it simply `myproject`.

**Comment by hombit** [comment]
Another idea to support @Hywan's initial proposal: one can rename Python module using simple macros based on cargo feature, i.e. we can have the same name for module and package for given feature if maturin will allow us to specify package name

**Comment by bovee** [comment]
I'm having the same issue where I have a rust library crate named e.g. "my_package" with Python bindings in a separate crate named "my_package_py", but I'd like the wheels to be named "my_package". In my case, I can't temporarily rename the bindings crate itself because it has a dependency on the library crate with the same name.

My super-hacky, temporary workaround was to clone maturin, change the [output name here](https://github.com/PyO3/maturin/blob/a1af8bf87be5890584e4b7c2a8670187fa59346f/src/metadata.rs#L143) to e.g. `"my_package".clone()` and then `cargo install --path .` it locally. Then I could build my wheels, rename them (removing the `_py`) and then upload them with e.g. `twine dist target/wheels/*`.

**Comment by tasn** [comment]
I have exactly the same use-case as @hombit as @bovee.
I understand the issue is that Python requires the binary name to be the same as the module name, and that's fine. One way to solve it would be to just let people wrap it in a python module, and inside it just have `from .lib_name import *`. Though this can't currently be done because maturin uses lib.name for the package name.

TL;DR: make it possible to tell maturin to use a directory (that's a valid python module) as the wheel's name.

**Comment by fjarri** [comment]
Blocked by this issue as well. I have a workspace with a Rust project `cratename` and Python bindings `cratename-python`, and would like the wheel (and hence the name in PyPi) to also be just `cratename`. I can't rename the bindings crate because it causes a conflict in the workspace. 

Would adding an optional `name` to `[package.metadata.maturin]` to override the wheel name work as a solution?

**Comment by Wenzel** [comment]
Hi,

I successfully built my first python bindings with PyO3 yesterday, and i was considering maturin for package and publication,
however, this issue is a blocker for me as well.

:arrow_right:  should i switch to `setuptools-rust` in the meantime ?
:arrow_right: do you intent to propose a fix to specify the package name ?

thanks !

**Comment by konstin** [comment]
What works thanks to houqp in #373 is the following, which should solve the `myproject` vs `myproject-python` case:

```toml
[package]
authors = ["konstin <konstin@mailbox.org>"]
name = "pyo3-name1"
version = "2.1.2"
edition = "2018"

[package.metadata.maturin]
name = "pyo3_name2"

[dependencies]
pyo3 = { version = "0.13.2", features = ["abi3-py36", "extension-module"] }

[lib]
name = "pyo3_name2"
crate-type = ["cdylib"]
```

This will create a `pyo3_name2` wheel.

> However, I want to generate multiple wheels (depending of some Cargo features), like myproject, myproject-xxx, myproject-yyy etc. Is it possible to change that from the maturin commands, or do I have to open a PR?

For this specific case, I suggest creating multiple wrapper crates in the same workspace that do nothing more than depending on the implementation with the right set of features and having the appropriate name.
